### PR TITLE
Unload passengers / cargo on death

### PR DESCRIPTION
**Pull request by teees**
Bit hacky approach, works by nudging passengers on exit so they don't bunch up.

The changes in EnterTransport will cancel any of the passengers activity, since mobile will discard nudges if not idle.

Kills any cargo it can't unload (when in water etc)

Edit:
No changes in EnterTransport, added a check in Enter to move on to the next activity and don't wait for the next tick. Also use Mobile.Nudge instead of INotifyBlockingMove.OnNotifyBlockingMove

http://gfycat.com/ImpishImaginativeFlea

https://gfycat.com/BoilingLimpAfricanjacana

### CODE REVIEW

File: mods/ts/rules/aircraft.yaml [hunk 1]
```diff
@@ -22,6 +22,7 @@ DPOD:
 		MaxWeight: 1
 		PipCount: 1
 		UnloadVoice: Move
+		EjectOnDie: true
```
Comment (GraionDilach) [inline mods/ts/rules/aircraft.yaml]: You refer this as `EjectOnDeath` in code. I prefer the code's form though.

Comment (teees) [inline mods/ts/rules/aircraft.yaml]: Just noticed that, refactored it before commit and forgot about it

File: OpenRA.Mods.Common/Traits/Cargo.cs [hunk 2]
```diff
@@ -300,6 +303,12 @@ public void Load(Actor self, Actor a)
 
 		public void Killed(Actor self, AttackInfo e)
 		{
+			if (Info.EjectOnDeath)
+			{
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Cargo.cs]: Style nit: these braces can go.

File: OpenRA.Mods.Common/Traits/Cargo.cs [hunk 3]
```diff
@@ -331,7 +340,9 @@ void SpawnPassenger(Actor passenger)
 				w.Add(passenger);
 				passenger.Trait<IPositionable>().SetPosition(passenger, self.Location);
 
-				// TODO: this won't work well for >1 actor as they should move towards the next enterable (sub) cell instead
+				var mobile = passenger.TraitOrDefault<Mobile>();
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Cargo.cs]: Can you use `OnNotifyBlockingMove` instead?  Hardcoding mobile references through unrelated code kind of sucks.

File: OpenRA.Mods.Common/Traits/Cargo.cs [hunk 4]
```diff
@@ -36,6 +36,9 @@ public class CargoInfo : ITraitInfo, Requires<IOccupySpaceInfo>
 		[Desc("When this actor is sold should all of its passengers be unloaded?")]
 		public readonly bool EjectOnSell = true;
 
+		[Desc("When this actor dies should all of its passengers be unloaded?")]
+		public readonly bool EjectOnDeath = false;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Cargo.cs]: It would be nice to have an option to eject the infantry with a health hit, but that makes this a bit more complicated so isn't required for my :+1:

**Comment by pchote** [comment]
Works well ingame and matches the original TS behaviour. :+1: after addressing the nits above.

**Comment by pchote** [comment]
Closes #3768.

**Comment by teees** [comment]
Changed and rebased

File: OpenRA.Mods.Common/Traits/Cargo.cs [hunk 5]
```diff
@@ -317,6 +320,10 @@ public void Load(Actor self, Actor a)
 
 		public void Killed(Actor self, AttackInfo e)
 		{
+			if (Info.EjectOnDeath)
+				while (!IsEmpty(self) && CanUnload())
```
Comment (obrakmann) [inline OpenRA.Mods.Common/Traits/Cargo.cs]: Is there a reason why you don't check for `DistanceAboveTerrain` here already? Then you wouldn't even need to run `SpawnPassengers`, including all the trait lookups and whatnot, in that situation.

File: OpenRA.Mods.Common/Traits/Cargo.cs [hunk 6]
```diff
@@ -343,13 +350,20 @@ public void Sold(Actor self)
...
-			});
+			if (positionable.CanEnterCell(self.Location, self, false) && !inAir)
+			{
+				self.World.AddFrameEndTask(w => w.Add(passenger));
+				var nbm = passenger.TraitOrDefault<INotifyBlockingMove>();
+				if (nbm != null)
+					nbm.OnNotifyBlockingMove(passenger, passenger);
+			}
+			else
+				passenger.Dispose(); // TODO: this may need to be a Kill() for statistics
```
Comment (obrakmann) [inline OpenRA.Mods.Common/Traits/Cargo.cs]: Yes, I think `Kill()` would be better.

File: OpenRA.Mods.Common/Traits/Cargo.cs (hunk 5 above)
Comment (teees) [inline OpenRA.Mods.Common/Traits/Cargo.cs]: The way it is now not really anymore. I initially wanted to support parachutes as well (see EjectOnDeath). Will change it

**Comment by Mailaender** [comment]
Also needs a rebase.

**Comment by teees** [comment]
rebased and adapted to the requested changes

**Comment by Mailaender** [comment]
Sorry, needs a rebase again.

**Comment by teees** [comment]
rebased

**Comment by Mailaender** [comment]
Both cases work fine. :white_check_mark:

**Comment by Mailaender** [comment]
Nicely done. Another tick on our list: https://github.com/OpenRA/OpenRA/issues/7874
